drivers:gpio_emul: add readback of output pins

evenl opened the pull request:
Currently the gpio_emul driver does not support reading back output pins through the port_get_raw function. It should as this is a common feature of drivers for true hardware (NRF and STM at least).

evenl commented:
I am not sure if this is the best solution, but it was the least intrusive. My first proposal was to remove the output_vals and input_vals members in the struct and only use a single member called values.

pdgendt commented:
@evenl We can't use the suggestion mechanism from github, you'll have to squash the commits locally and force push a single commit.

evenl commented:
Yes, I know, I was just lazy. It was easier for me to just commit your patch, and then squash it afterwards.

cfriedt reviewed (changes requested):
The API specifically says input pins. 

https://docs.zephyrproject.org/latest/hardware/peripherals/gpio.html#c.gpio_port_get_raw

However, there is nothing preventing a gpio pin from being both an input and an output.

If your output pins are also configured as input, does this work as you expect?

Please include some test coverage if none exists already

evenl commented:
> However, there is nothing preventing a gpio pin from being both an input and an output.
> If your output pins are also configured as input, does this work as you expect?
> Please include some test coverage if none exists already

If that is the case the STM and NRF gpio drivers must be updated, because they do not honor that behavior. I have not tested if it works as expected if the pin is both input and output (actually I have to admit that I assumed that was mutual exclusive). I'll get back to you when I have tested.

cfriedt commented:
> > However, there is nothing preventing a gpio pin from being both an input and an output.
> > If your output pins are also configured as input, does this work as you expect?
> > Please include some test coverage if none exists already
> 
> If that is the case the STM and NRF gpio drivers must be updated, because they do not honor that behavior. I have not tested if it works as expected if the pin is both input and output (actually I have to admit that I assumed that was mutual exclusive). I'll get back to you when I have tested.

it's quite possible that those platforms may be hard-wired to always have the input channel read-back the output channel. Different GPIO controllers can very well behave differently.

@nordicjm or @erwango - do you know if that's the case?

nordicjm commented:
https://infocenter.nordicsemi.com/pdf/nRF52840_PS_v1.7.pdf on page 153 shows the GPIO architecture, there is an enable for the input to be connected or disconnected, when connected it can back the state of an output, yes. From the HAL perspective directly (ignoring zephyr) an output pin can be read without making it an input.

erwango commented:
> @nordicjm or @erwango - do you know if that's the case?

H742 RM specifialy mentions "When the I/O port is programmed as output: [...] A read access to the input data register gets the I/O state" (didn't check all RM, but there are very few variations on GPIO HW block).

evenl commented:
Seems that setting the pins as both output and input is working, but the callbacks on output pins stops working. The reason is this line:

`gpio_fire_callbacks(&drv_data->callbacks, port, pins & ~get_input_pins(port));`

in function: `static int gpio_emul_port_set_bits_raw(const struct device *port, gpio_port_pins_t pins) `

I changed the line to this:

`gpio_fire_callbacks(&drv_data->callbacks, port, pins & get_output_pins(port));`

Then it works. I do not understand why it's written like that? It's done like that for all of the output functions like port_set_bits_raw, port_set_masked_raw, port_clear_bits_raw and port_toggle_bits.

I can update this PR with this change, or do you prefer that I open a new PR? The title is kind of wrong now :)

Commit:
drivers: gpio_emul: output-input gpio pins does not trigger callback

If a pin is set to both input and output the pin will not trigger the
callback when  the value is changed.

Signed-off-by: Even Falch-Larsen <even.falch.larsen@nomono.co>

henrikbrixandersen commented:
> I can update this PR with this change, or do you prefer that I open a new PR? The title is kind of wrong now :)

Please open a separate PR for this.
